### ISSUE TITLE
Messages not being displayed

### ISSUE THREAD

**Issue by Moffinguer** [bmeneg/coc-perl#19]
Hello, I have just installed this extension since I'm moving from VSCode to NVim.

One of the problems I'm encountering with is that the error messages that should be displayed by line, are not being shown, even if my mouse or cursor is on the line. This only appears on this extension, but not on others.

![imagen](https://user-images.githubusercontent.com/55874386/235806753-c7d43d5b-6413-4d06-8855-1b6b359310e3.png)

![imagen](https://user-images.githubusercontent.com/55874386/235806810-405850ec-f820-48c3-96c8-4fe942a1a455.png)

I have no aditional configuration. Is there any more data I should provide?

**Comment by bmeneg** [comment]
@Moffinguer I'm really sorry for the late reply.

The last time I checked it was working, something might have changed on `coc-nvim` to make it just stop.
I'll take a look on that and fix whatever is needed.

**Comment by Moffinguer** [comment]
Oh @bmeneg it's all right!, since it was my first time dealing with LSP with coc, i thought it was my fault, so i simply used a vanilla solution meanwhile.

**Comment by bmeneg** [comment]
But it's indeed not working as it should.
I may have some spare time to check it tomorrow, hopefully I can find the culprit of it, but it's, probably, outside `coc-perl` scope, meaning I'll have to dive into [language server side](https://github.com/richterger/Perl-LanguageServer).

**Comment by bmeneg** [comment]
Yeap, that's indeed an error on server side. The server sends two notification to the client about "syntax check": one it says there are errors, while the other notification says "things are fine". I'm not sure why, but VSCode seems to know how to handle this "difference" (stacking notifications?), while coc.nvim just overrides the first notification by the second, so depending on the order the notifications are sent, the pop-up doesn't appears or other strange things starts to happen.

Although I'm not sure how VSCode vs coc-nvim work behind the scenes for this case, the Perl server is broken on sending two notifications for the same "issue" and confusing things, thus I'm going to fix the server and get the same behavior in both clients (VSCode and coc-nvim).

**Comment by bmeneg** [comment]
@bukephalos just reported (issue #18) the exact behavior I saw during this issue investigation yesterday:

> * Install your most recent version. When the install was done everything looked very good. Errors were marked and the corresponding Perl error messages shown below.
> * Then I closed vim and opened the same file again. Now the errors were still marked but the messages are no longer shown.
> * After more open/close cycles or creating a new file also with a syntax error nothing is shown. Just the Perl code but no error indicator. CocInfo gives no error and lookes as if the Language Server was started/connected. But as I said, no noticable effect.
> * I did a reboot and was back to the second state: error indicators (two ancle brackets to the left of the lines with syntax errors) but no messages about the type of error.
> * I fixed one error (the file contained several syntax errors) and all error indicators disappeared. Very strange.

While testing, I also notice that VSCode is also having the very same issue, which yesterday it didn't.
Basically, this inconsistency is due to the ordering by which each notification is sent to the client (VSCode or coc.nvim). The structure holding the notifications in the server code is a dictionary, which isn't ordered and every `for`-loop on it the order might be different. With that, when the "wrong" notification is sent last, it overrides the "correct" notification with real "syntax check" problems, confusing the clients.

I already have a possible fix for the problem on server code, I'll be posting it in a few and hopefully we can get it merged during the week.

**Comment by bmeneg** [comment]
Simple PR submitted to server: richterger/Perl-LanguageServer/pull/185. 
As soon as it get fixed I'll ask the maintainer to generate a hotfix release.

**Comment by bukephalos** [comment]
Thanks for the fix. Looks better with it but it still doen't work as expected. I guess it is another problem with LanguageServer or with my understanding. Doesn't really belong here but I am not sure where else to ask (is there a Perl-LanguageServer discussion forum/mailing list or something similar?).

I have a very simple test script:
```
#!/usr/bin/perl

use strict;
use warnings;

$a = 112;

dosomething();

sub dosomething {
    print "foo\n";
}
```

It doesn't indicate the missing "my" before the variable definition. But if I remove the semicolon from the end of the definition the correct diagnostics is shown (missing simcolon and syntax error). I remember that in my tests I have seen a "global symbol..." diagnostic indicating a missing "my" but not in this little example.

Second problem/misunderstanding: I move the cursor to "dosomething()" and type "gd" to go to the sub definition but I only get the message "[coc.nvim] definition not found".
As I said, it might be a misunderstanding on my part but I am just trying to move from simple editor to IDE functionality use and see that I have a lot to learn (any idea where to get help with that?).

**Comment by bmeneg** [comment]
@bukephalos feel free to ask here :) 

For the first question, related to `$a` assignment, no errors is indeed to correct behavior: I have no idea why Perl developers it would be a good idea, but both `$a` and `$b` are always globally declared by the interpreter to be used with the `sort()` function whenever a code block is passed to it like `sort({$a > $b})`, you basically can't lexically declare them on your package like `my $a; my $b` (for more info: https://perldoc.perl.org/perlvar#$a). If you change the `$a` to `$c` you're going to see the error message.

For the second question: how are you configuring coc-perl on CoC config (`:CocConfig`)? Is it being added within a `"languageserver"` object is it by its own object starting with `"perl"`? Do you have the following?
```json
{
    "perl": {
        "enable": true
    }
}
```
How is your `coc-definition` (`gd`) defined at your vimrc/init.vim, is it like the following?
```vim
nmap <silent> gd <Plug>(coc-definition)
```

**Comment by bukephalos** [comment]
Thanks for your help!

$a is understood now.

This is my CocConfig:
```
{
      "languageserver": {
         "perl": {
          "enable": true,
          "filetypes": ["perl"]
        }
      }
  }
```
gives the warning/error message:
   Missing property "module". (json)

once I also got this message when I opened a perl file:
   [coc.nvim] Wrong configuration of LS "perl", no command or module specified.

But seems to work at least for syntax checks but still no luck with gd. My key mapping is exactly as yours and it seems to be unterstood, at least I get the message:
   [coc.nvim] definition not found
which indicates it is looking for a definition.

And I get the same ("no definition found for 'dosomething'") in VScode with the same little example script I quoted yesterday.
This is why I suspect that it is a misunderstanding on my side.

Does it work for you? With the same script, same CocConfig and same key binding? Does it perhaps only work for methods and not for simple subs? A test within one of the Perl::LanguageServer source files worked (somewhat): didn't go to the definition but showed the source of the definition near my cursor. But I have no idea what conditions have to be met for this functionality to work (should it also work for variable definitions? Didn't for me)

**Comment by bmeneg** [comment]
@bukephalos, ah I see now what's going on, you're using the coc-perl configuration from within the `"languageserver"` property. This extension property is for configuring VSCode internal language server client support, which isn't what want. Since you're using coc-perl you need to call it directly from you `coc-settings.json` file, `:CocConfig`, (or `settings.json` if using VSCode):
```
{
    "perl": {
        "enable": true
    }
}
```
You don't place the above inside an `"languageserver": {}`.
Also, make sure to remove the `fileTypes: []` property, since it defines the file type Perl::LanguageServer should check for Perl codes, which, but default, are set to `[ ".pl", ".pm" ]` by coc-perl itself.

**Comment by bukephalos** [comment]
Followed your suggestion but still the same: gd gives "[coc.nvim] definition not found"

CocConfig looks exactly like yours:
```
{
⚠     "perl": {
         Property perl is not allowed. (json) 
      }
  }
```
And as you can see coc-json shows a syntax error with it. (the diagnostics hides the "enable": true which is also present)
To be sure, here it is as a complete cut and paste without the diagnostic message:
```
{
     "perl": {
          "enable": true
      }
}
```

**Comment by bmeneg** [comment]
@bukephalos, the json diagnostic about perl not being allowed is fine, it won't affect the overall functionality (although I still need to find what's causing it). But the `[coc.nvim] definition not found` is strange. I tested with your example and `gd` just worked. Would you mind opening a new bug report so we can dive into this specifically?

In the new issue, could you add the following configurations to your coc settings?
```
{
  "perl": {
    "enable": true,
    "logFile": "<any-folder>/coc-perl.log",
    "logLevel": 2
  }
}
```
1. Make sure to substitute `<any-folder>` by a location in your machine where you can open later;
2. close every Perl file you've opened with vim/neovim; 
3. open only your example Perl file
4. run on your neovim the command `:CocOpenLog`, save it to a file and paste in the issue;
5. open the `coc-perl.log` that you registered in the coc-settings.json and paste its content to https://l.perl.bot/ (click on "submit" button under the text field after you've pasted all log content), then copy and paste the URL in the new issue.

**Comment by bmeneg** [comment]
> Simple PR submitted to server: https://github.com/richterger/Perl-LanguageServer/pull/185.
> As soon as it get fixed I'll ask the maintainer to generate a hotfix release.

A new version of the server was released containing the fix for the issue originally pointed in this issue.
I just released v1.3.1 which include the fix. Just make sure to `cpan Perl::LanguageServer` to update the server in your local machine.
